File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;

        for(char c: s) {
            if(c == '[' || c == '{' || c == '(') {
                stk.push(c);
            } else if(c == ']') {
                if(stk.empty()) return false;
                if(stk.top() == '[') {
                    stk.pop();
                } else {
                    return false;
                }
            } else if(c == '}') {
                if(stk.empty()) return false;
                if(stk.top() == '{') {
                    stk.pop();
                } else {
                    return false;
                }
            } else if(c == ')') {
                if(stk.empty()) return false;
                if(stk.top() == '(') {
                    stk.pop();
                } else {
                    return false;
                }
            }
        }

        return stk.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()

        for c in s:
            if c == '(' or c == '[' or c == '{':
                stack.append(c)
            else:
                if len(stack) == 0:
                    return False
                elif c == ')':
                    if stack[-1] != '(':
                        return False
                    stack.pop()
                elif c == ']':
                    if stack[-1] != '[':
                        return False
                    stack.pop()
                elif c == '}':
                    if stack[-1] != '{':
                        return False
                    stack.pop()
        
        return len(stack) == 0